### PR TITLE
ARROW-7858: [C++][Python] Support casting from ExtensionArray

### PR DESCRIPTION
**Pull request by kszucs**
This add support for casting from ExtensionType. We should probably add support for converting to an ExtensionType, but we can handle it in a follow-up.

### CODE REVIEW

**Comment by kszucs** [comment]
The test failure is related to #6632

**Review by jorisvandenbossche — COMMENTED** [review]
Casting to its own storage type (so eg not a different integer bit), still returns an ExtensionArray. With the example from the test:

```
In [3]: arr.cast(pa.int64()) 
Out[3]: 
<pyarrow.lib.ExtensionArray object at 0x7f4b45ca5fa8>
[
  1,
  2,
  3,
  4
]

In [4]: arr.cast(pa.int32()) 
Out[4]: 
<pyarrow.lib.Int32Array object at 0x7f4b45ca5348>
[
  1,
  2,
  3,
  4
]
```

I am not fully sure that this is the expected behaviour.

**Comment by kszucs** [comment]
Nice catch, it was caused by the identity shortcut. Added a separate kernel which handles the identity case properly.

**Review by jorisvandenbossche — APPROVED** [review]
Looks good to me now! (but not very familiar with the cast kernel code)

File: cpp/src/arrow/compute/kernels/cast.cc [hunk 1]
```diff
@@ -1127,6 +1127,42 @@ class ZeroCopyCast : public CastKernelBase {
...
+    kernel->reset(
+        new ExtensionCastKernel(std::move(storage_caster), std::move(out_type)));
+
+    return Status::OK();
+  }
+
+  Status Call(FunctionContext* ctx, const Datum& input, Datum* out) override {
+    DCHECK_EQ(input.kind(), Datum::ARRAY);
+    if (input.type()->id() != Type::EXTENSION) {
+      return Status::TypeError("Not an extension type");
```
Comment (bkietz) [inline cpp/src/arrow/compute/kernels/cast.cc]: This should probably also verify that the `storage_type` (and maybe the `extension_name`) is the same as the input type the kernel was constructed with.

**Review by bkietz — CHANGES_REQUESTED** [review]
Looks good, just a few notes

File: python/pyarrow/tests/test_extension_type.py [hunk 2]
```diff
@@ -168,6 +177,37 @@ def test_ext_array_pickling():
...
+        assert isinstance(casted, klass)
+
+    # test chunked array casting
+    arr = pa.chunked_array([arr, arr])
+    casted = arr.cast(pa.int16())
+    assert casted.type == pa.int16()
+    assert isinstance(casted, pa.ChunkedArray)
+
+
+def test_casting_to_extension_type_raises():
```
[the remark this answers is not in the document]
Comment (kszucs) [reply python/pyarrow/tests/test_extension_type.py]: Added.

File: cpp/src/arrow/compute/kernels/cast.cc [hunk 3]
```diff
@@ -1127,6 +1127,62 @@ class ZeroCopyCast : public CastKernelBase {
...
+    }
+
+    // validate: type is the same as the type the kernel was constructed with
+    auto& input_type = checked_cast<const ExtensionType&>(*input.type());
+    if (input_type.extension_name() != extension_name_) {
+      return Status::TypeError("EEEE");
```
Comment (kszucs) [inline cpp/src/arrow/compute/kernels/cast.cc]: Add meaningful error messages.

File: cpp/src/arrow/compute/kernels/cast.cc (hunk 1 above)
Comment (kszucs) [reply cpp/src/arrow/compute/kernels/cast.cc]: Updated.

**Comment by kszucs** [comment]
@bkietz please take a look again

File: cpp/src/arrow/compute/kernels/cast.cc [hunk 4]
```diff
@@ -1127,6 +1127,64 @@ class ZeroCopyCast : public CastKernelBase {
...
+  }
+
+  Status Call(FunctionContext* ctx, const Datum& input, Datum* out) override {
+    DCHECK_EQ(input.kind(), Datum::ARRAY);
+    if (input.type()->id() != Type::EXTENSION) {
+      return Status::TypeError("Not an extension type");
+    }
+
+    // validate: type is the same as the type the kernel was constructed with
+    auto& input_type = checked_cast<const ExtensionType&>(*input.type());
```
Comment (bkietz) [inline cpp/src/arrow/compute/kernels/cast.cc]:
```suggestion
    const auto& input_type = checked_cast<const ExtensionType&>(*input.type());
```

File: cpp/src/arrow/compute/kernels/cast.cc [hunk 5]
```diff
@@ -1127,6 +1127,64 @@ class ZeroCopyCast : public CastKernelBase {
...
+    if (input_type.extension_name() != extension_name_) {
+      return Status::TypeError(
+          "The cast kernel was constructed with a differently named extension type");
```
Comment (bkietz) [inline cpp/src/arrow/compute/kernels/cast.cc]: This error message (and the one below for storage_type) can be more helpful:
```suggestion
          "The cast kernel was constructed to cast from the extension type named '", extension_name_, 
          "' but input has named extension type name '", input_type.extension_name(), "'");
```

File: cpp/src/arrow/compute/kernels/cast_test.cc [hunk 6]
```diff
@@ -1480,5 +1482,50 @@ TYPED_TEST(TestDictionaryCast, OutTypeError) {
...
 }*/
 
+std::shared_ptr<Array> SmallintArrayFromJSON(const std::string& json_data) {
+  auto arr = ArrayFromJSON(int16(), json_data);
+  auto ext_data = arr->data()->Copy();
+  ext_data->type = smallint();
+  return MakeArray(ext_data);
+}
+
+TEST_F(TestCast, ExtensionTypeToIntDowncast) {
```
Comment (bkietz) [inline cpp/src/arrow/compute/kernels/cast_test.cc]: :+1:

**Review by bkietz — CHANGES_REQUESTED** [review]

File: cpp/src/arrow/compute/kernels/cast.cc (hunk 5 above)
Comment (kszucs) [reply cpp/src/arrow/compute/kernels/cast.cc]: Indeed.

Comment (bkietz) [reply cpp/src/arrow/compute/kernels/cast.cc]: please also make the error for storage type verbose

File: cpp/src/arrow/compute/kernels/cast.cc [hunk 7]
```diff
@@ -1127,6 +1127,66 @@ class ZeroCopyCast : public CastKernelBase {
...
+    }
+    if (!input_type.storage_type()->Equals(storage_type_)) {
+      return Status::TypeError(
+          "The cast kernel was constructed with a different extension type");
+    }
+
+    // construct an ArrayData object with the underlying storage type
+    auto new_input = input.array()->Copy();
+    new_input->type = storage_type_;
+    return InvokeWithAllocation(ctx, storage_caster_.get(), new_input, out);
```
Comment (wesm) [inline cpp/src/arrow/compute/kernels/cast.cc]: Does it allocate if the out_type and storage_type are the same?

File: cpp/src/arrow/compute/kernels/cast.cc [hunk 8]
```diff
@@ -1127,6 +1127,66 @@ class ZeroCopyCast : public CastKernelBase {
...
 };
 
+class ExtensionCastKernel : public CastKernelBase {
+ public:
+  static Status Make(const DataType& in_type, std::shared_ptr<DataType> out_type,
+                     const CastOptions& options,
+                     std::unique_ptr<CastKernelBase>* kernel) {
+    if (in_type.id() != Type::EXTENSION) {
+      return Status::TypeError("Not an extension type");
+    }
```
Comment (wesm) [inline cpp/src/arrow/compute/kernels/cast.cc]: I don't think this check is needed if this "can't happen" -- the checked_cast should be sufficient

File: cpp/src/arrow/testing/gtest_util.cc [hunk 9]
```diff
@@ -382,11 +382,7 @@ void SleepFor(double seconds) {
 // Extension types
 
 bool UUIDType::ExtensionEquals(const ExtensionType& other) const {
-  const auto& other_ext = static_cast<const ExtensionType&>(other);
-  if (other_ext.extension_name() != this->extension_name()) {
-    return false;
-  }
-  return true;
+  return (other.extension_name() == this->extension_name());
 }
```
Comment (wesm) [inline cpp/src/arrow/testing/gtest_util.cc]: I wonder if this should be the default implementation of `ExtensionEquals`

**Review by wesm — COMMENTED** [review]
The PR title misled me when I reviews this, it should be "Support casting the storage of an ExtensionArray" or similar

**Comment by wesm** [comment]
There was more C++ code in the implementation than I expected, not sure what could be done to make implementing something that "seems easy" (on paper) easier

File: cpp/src/arrow/compute/kernels/cast.cc (hunk 7 above)
Comment (kszucs) [reply cpp/src/arrow/compute/kernels/cast.cc]: According to the tests no allocation happens.

File: cpp/src/arrow/compute/kernels/cast.cc (hunk 8 above)
Comment (kszucs) [reply cpp/src/arrow/compute/kernels/cast.cc]: Removed

File: cpp/src/arrow/testing/gtest_util.cc (hunk 9 above)
Comment (kszucs) [reply cpp/src/arrow/testing/gtest_util.cc]: Created a follow-up JIRA https://issues.apache.org/jira/browse/ARROW-8143

**Review by wesm — APPROVED** [review]
+1. Thanks @kszucs

**Pull request closed without merging**